File: apps/frontend/src/adapters/shared/provenance.ts
import type { ChatSourceEmail } from "@/lib/types";
import { invoke, logger } from "./platform";

export const listSourceEmails = async (params?: {
  threadId?: string;
  limit?: number;
}): Promise<ChatSourceEmail[]> => {
  try {
    return await invoke<ChatSourceEmail[]>("list_source_emails", params ?? {});
  } catch (error) {
    logger.error("Error listing source emails.");
    throw error;
  }
};

File: apps/server/src/api/provenance.rs
use std::sync::Arc;

use crate::{error::ApiResult, main_lib::AppState};
use axum::{
    extract::{Query, State},
    routing::get,
    Json, Router,
};
use serde::Deserialize;
use wealthfolio_core::provenance::ChatSourceEmail;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct EmailQuery {
    #[serde(default)]
    thread_id: Option<String>,
    #[serde(default)]
    limit: Option<i64>,
}

async fn list_source_emails(
    Query(q): Query<EmailQuery>,
    State(state): State<Arc<AppState>>,
) -> ApiResult<Json<Vec<ChatSourceEmail>>> {
    let emails = match q.thread_id {
        Some(thread_id) => state.provenance_service.thread_emails(&thread_id).await?,
        None => {
            state
                .provenance_service
                .recent_emails(q.limit.unwrap_or(100))
                .await?
        }
    };
    Ok(Json(emails))
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new().route("/provenance/source-emails", get(list_source_emails))
}
